Sources/PostgresNIO/New/PSQLRowSequence.swift

```diff
@@ -0,0 +1,607 @@
+        }
+        let iterator = PSQLRowSequence.Iterator(consumer: self)
+        return iterator
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.lock.withLock {
+            self.state.iteratorDeinitialized()
+        }
+
+        switch action {
+        case .cancelStream(let source):
+            source.cancel()
+        case .none:
+            break
+        }
+    }
+    
+    func next() async throws -> PSQLRow? {
+        self.lock.lock()
+        switch self.state.next() {
+        case .returnNil:
+            self.lock.unlock()
+            return nil
+            
+        case .returnRow(let data, signalDemandTo: let source):
+            self.lock.unlock()
+            source?.demand()
+            return PSQLRow(
+                data: data,
+                lookupTable: self.lookupTable,
+                columns: self.columns,
+                jsonDecoder: self.jsonDecoder
+            )
+            
+        case .throwError(let error):
+            self.lock.unlock()
+            throw error
+            
+        case .hitSlowPath:
+            return try await withCheckedThrowingContinuation { continuation in
+                let slowPathAction = self.state.next(for: continuation)
+                self.lock.unlock()
+                switch slowPathAction {
+                case .signalDemand(let source):
+                    source.demand()
+                case .none:
+                    break
+                }
+            }
+        }
+    }
+    
+}
+
+extension AsyncStreamConsumer {
+    struct StateMachine {
+        enum UpstreamState {
+            enum DemandState {
+                case canAskForMore
+                case waitingForMore(CheckedContinuation<PSQLRow?, Error>?)
+            }
+            
+            case initialized
+            case streaming(AdaptiveRowBuffer, PSQLRowStream, DemandState)
+            /// The upstream has finished, but the downstream has not consumed all events.
+            case finished(AdaptiveRowBuffer, String)
+            /// The upstream has failed, but the downstream has not consumed the error yet.
+            case failed(Error)
+            /// The upstream has failed or finished and the downstream has consumed all events. Final state.
+            case consumed
+
+            case modifying
+        }
+        
+        enum DownstreamState {
+            case sequenceCreated
+            case iteratorCreated
+        }
+        
+        var upstreamState: UpstreamState
+        var downstreamState: DownstreamState
+        
+        init() {
+            self.upstreamState = .initialized
+            self.downstreamState = .sequenceCreated
+        }
+        
+        mutating func buffered(_ buffer: CircularBuffer<DataRow>, upstream: PSQLRowStream) {
+            guard case .initialized = self.upstreamState else {
+                preconditionFailure("Invalid upstream state: \(self.upstreamState)")
+            }
```

Comment:
```suggestion
            precondition(self.upstreamState == .initialized, "Invalid upstream state: \(self.upstreamState)")
```

Reply:
As this is mostly a style nit, how serious are you about it?

Reply:
I tend to consider `assertionFailure()` and `preconditionFailure()` to be there as fallbacks: "if for whatever reason your condition can't be expressed in an autoclosure, here's your escape hatch". I make a point of only using them if the check _can't_ be written as an inline boolean expression (or if doing so would be impractical, like nested ternaries or such). Aside the conceptual reason, IMO it makes it easier to tell the difference between enforcing invariants and normal code flow.

In this particular case, there's also the bloody awful backwards-arse `case .enumerator = value` syntax to reckon with - it has its place, primarily when dealing with associated values, but for trivial [in]equality enum comparisons I consider `Equatable` conformance to be a divine gift (yeah, I'm just a touch opinionated about the `case let` syntax, does it show? 😅). It especially bugs me when it becomes a reason for being forced to use `preconditionFailure()` 😆 

To answer your question, I mean, yeah, you're right, it's a style nit, I imagine it probably compiles down to the same IR in the end (wouldn't be surprised if it was a 1:1 correspondence instruction for instruction in fact) - but it hits on two significant questions of code readability and conceptual consistency simultaneously. I'm not gonna die on the hill, but this is one I definitely take seriously.